File: Plot Curve/plot_curve.c
#include <stdio.h>
#include <math.h>

#define WIDTH 80  // Width of the plot area
#define HEIGHT 20 // Height of the plot area

// Function definition: f(x) = x^3 + x + 1
double f(double x) {
    return pow(x, 3) + x + 1;
}

// Map function value to plot area height
int mapToHeight(double y, double minY, double maxY) {
    return (int)((y - minY) / (maxY - minY) * (HEIGHT - 1));
}

int main() {
    int num_segments = 20;  // Number of segments
    double a = -1;          // Start of the interval
    double b = 0;           // End of the interval
    double step = (b - a) / num_segments;  // Step size for the interval

    // Calculate min and max of function values for scaling
    double minY = f(a);
    double maxY = f(a);

    for (int i = 1; i <= num_segments; i++) {
        double x = a + i * step;   // Calculate x value
        double y = f(x);          // Calculate function value

        if (y < minY) minY = y;
        if (y > maxY) maxY = y;
    }

    // Create the plot area
    char plot[HEIGHT][WIDTH];
    for (int i = 0; i < HEIGHT; i++) {
        for (int j = 0; j < WIDTH; j++) {
            plot[i][j] = ' ';
        }
    }

    // Draw the line segments
    for (int i = 0; i < num_segments; i++) {
        double x1 = a + i * step;
        double x2 = a + (i + 1) * step;
        double y1 = f(x1);
        double y2 = f(x2);

        int x1_plot = (int)((x1 - a) / (b - a) * (WIDTH - 1));
        int x2_plot = (int)((x2 - a) / (b - a) * (WIDTH - 1));
        int y1_plot = mapToHeight(y1, minY, maxY);
        int y2_plot = mapToHeight(y2, minY, maxY);

        if (x1_plot != x2_plot) {
            int x_start = x1_plot < x2_plot ? x1_plot : x2_plot;
            int x_end = x1_plot > x2_plot ? x1_plot : x2_plot;

            for (int x = x_start; x <= x_end; x++) {
                int y_plot = y1_plot + (y2_plot - y1_plot) * (x - x1_plot) / (x2_plot - x1_plot);
                if (y_plot >= 0 && y_plot < HEIGHT) {
                    plot[HEIGHT - 1 - y_plot][x] = '*';
                }
            }
        }
    }

    // Print the plot
    for (int i = 0; i < HEIGHT; i++) {
        for (int j = 0; j < WIDTH; j++) {
            printf("%c", plot[i][j]);
        }
        printf("\n");
    }

    return 0;
}

File: Plot Curve/plot_curve.py
import numpy as np

# Function definition: f(x) = x^3 + x + 1
def f(x):
    return x**3 + x + 1

# Parameters
num_segments = 20  # Number of segments
a = -1             # Start of the interval
b = 0              # End of the interval
width = 80         # Width of the plot area
height = 20        # Height of the plot area

# Calculate step size
step = (b - a) / num_segments

# Create an empty plot
plot = [[' ' for _ in range(width)] for _ in range(height)]

# Calculate min and max values of the function for scaling
x_values = np.linspace(a, b, num_segments + 1)
y_values = [f(x) for x in x_values]
min_y = min(y_values)
max_y = max(y_values)

# Map function value to plot area height
def map_to_height(y, min_y, max_y):
    return int((y - min_y) / (max_y - min_y) * (height - 1))

# Draw the line segments
for i in range(num_segments):
    x1 = a + i * step
    x2 = a + (i + 1) * step
    y1 = f(x1)
    y2 = f(x2)

    x1_plot = int((x1 - a) / (b - a) * (width - 1))
    x2_plot = int((x2 - a) / (b - a) * (width - 1))
    y1_plot = map_to_height(y1, min_y, max_y)
    y2_plot = map_to_height(y2, min_y, max_y)

    if x1_plot != x2_plot:
        for x in range(min(x1_plot, x2_plot), max(x1_plot, x2_plot) + 1):
            y_plot = y1_plot + (y2_plot - y1_plot) * (x - x1_plot) / (x2_plot - x1_plot)
            if 0 <= y_plot < height:
                plot[height - 1 - int(y_plot)][x] = '*'

# Print the plot
for row in plot:
    print(''.join(row))
